AiSceneLoad does not load array type user parameter from usd

**Describe the bug**
The attached archive has the test code.
The test code does the following:
create a node
declare a user array parameter on the node
save the node to .usd
load the node from .usd
verify if the node has the user array parameter set
The test fails with the following error:
WARNING | myimager: could not read ARRAY parameter "my_array"
[ERROR] 'my_array' is defined null 
Changing the format to .ass works fine.
Writing to usda and looking at the file, I can tell that the array is written correctly, so it's likely an issue with the loader:
    float[] primvars:my_array = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15] (
        elementSize = 16
        interpolation = "constant"
    )

[UsdUserArrayTest.cpp.zip](https://github.com/Autodesk/arnold-usd/files/9707554/UsdUserArrayTest.cpp.zip)

--- a/testsuite/test_1315/README
+++ b/testsuite/test_1315/README
@@ -1,0 +1,5 @@
+AiSceneLoad does not load array type user parameter from usd
+
+see #1315
+
+author: Cyril Pichard

--- a/testsuite/test_1315/data/test.cpp
+++ b/testsuite/test_1315/data/test.cpp
@@ -1,0 +1,96 @@
+#include <ai.h>
+
+#include <stdio.h>
+
+#define SCENE_FILE "scene.usda"
+
+int save_scene()
+{
+    int errorCode = 0;
+    AtUniverse* universe = AiUniverse();
+    /* create node */
+    AtNode* imager = AiNode(universe, AtString("imager_color_curves"), AtString("myimager"));
+    /* declare user array parameter */
+    AiNodeDeclare(imager, AtString("my_array"), AtString("constant ARRAY FLOAT")); // Matrice ????
+    /* set user array parameter */
+    AtArray* values = AiArray(16, 1, AI_TYPE_FLOAT);
+    for (int i = 0; i < 16; i++)
+        AiArraySetFlt(values, i, (float)i);
+    AiNodeSetArray(imager, AtString("my_array"), values);
+
+    /* save to file */
+    AtParamValueMap* params = AiParamValueMap();
+    if (!AiSceneWrite(universe, SCENE_FILE, params, NULL)) {
+        printf("[ERROR] Unable to write " SCENE_FILE "\n");
+        errorCode = 1;
+    }
+    AiParamValueMapDestroy(params);
+
+    AiUniverseDestroy(universe);
+    return errorCode;
+}
+
+int load_scene()
+{
+    int errorCode = 0;
+    AtUniverse* universe = AiUniverse();
+    AtNode* imager = NULL;
+    AtArray* values = NULL;
+
+    /* load from file */
+    AtParamValueMap* params = AiParamValueMap();
+    if (!AiSceneLoad(universe, SCENE_FILE, params)) {
+        printf("[ERROR] Failed to read " SCENE_FILE "\n");
+        errorCode = 1;
+        AiParamValueMapDestroy(params);
+        goto release_universe;
+    }
+    AiSceneWrite(universe, "scene.ass", params, NULL);
+    AiParamValueMapDestroy(params);
+
+    /* verify */
+    imager = AiNodeLookUpByName(universe, AtString("myimager"));
+    if (imager == nullptr) {
+        printf("[ERROR] 'myimager' nodes is not found\n");
+        errorCode = 1;
+        goto release_universe;
+    }
+    if (AiNodeLookUpUserParameter(imager, AtString("my_array")) == nullptr) {
+        printf("[ERROR] 'my_array' user parameter is not found\n");
+        errorCode = 1;
+        goto release_universe;
+    }
+    values = AiNodeGetArray(imager, AtString("my_array"));
+    if (values == nullptr) {
+        printf("[ERROR] 'my_array' is defined null\n");
+        errorCode = 1;
+        goto release_universe;
+    }
+    if (AiArrayGetNumElements(values) != 16) {
+        printf("[ERROR] 'my_array' has incorrect number of elements: %u\n", AiArrayGetNumElements(values));
+        errorCode = 1;
+        goto release_universe;
+    }
+    for (int i = 0; i < 16; i++) {
+        float v = AiArrayGetFlt(values, i);
+        if (v != (float)i) {
+            printf("[ERROR] 'my_array' %d. element is incorrect: %f\n", v);
+            errorCode = 1;
+            goto release_universe;
+        }
+    }
+release_universe:
+    AiUniverseDestroy(universe);
+    return errorCode;
+}
+
+int main(int argc, const char** argv)
+{
+    int errorCode = 0;
+    AiBegin(AI_SESSION_BATCH);
+    AiMsgSetConsoleFlags(nullptr, AI_LOG_WARNINGS | AI_LOG_ERRORS | AI_LOG_BACKTRACE);
+    errorCode += save_scene();
+    errorCode += load_scene();
+    AiEnd();
+    return errorCode;
+}

--- a/translator/reader/prim_reader.cpp
+++ b/translator/reader/prim_reader.cpp
@@ -654,10 +654,21 @@ void UsdArnoldPrimReader::ReadPrimvars(
 
         if (primvarType == AI_TYPE_NONE)
             continue;
-        declaration += " ";
-        declaration += AiParamGetTypeName(primvarType);
+
+        // Deduce primvar type and array type.
+        int arrayType = AI_TYPE_NONE;
+        if ((interpolation != UsdGeomTokens->constant && primvarType != AI_TYPE_ARRAY)
+        || (interpolation == UsdGeomTokens->constant && typeName.IsArray()) ) {
+            arrayType = primvarType;
+            primvarType = AI_TYPE_ARRAY;
+        }
 
         // Declare the user data
+        if (arrayType != AI_TYPE_NONE) {
+            declaration += " ARRAY";
+        }
+        declaration += " ";
+        declaration += AiParamGetTypeName(primvarType);
         AtString nameStr(name.GetText());
         if (AiNodeLookUpUserParameter(node, nameStr) == nullptr && 
             AiNodeEntryLookUpParameter(nodeEntry, nameStr) == nullptr) {
@@ -707,12 +718,6 @@ void UsdArnoldPrimReader::ReadPrimvars(
 
                 hasIdxs = true;
             }
-        }
-
-        int arrayType = AI_TYPE_NONE;
-        if (interpolation != UsdGeomTokens->constant && primvarType != AI_TYPE_ARRAY) {
-            arrayType = primvarType;
-            primvarType = AI_TYPE_ARRAY;
         }
 
         InputAttribute inputAttr(primvar);

